Gem/Asset/AssetScene.swift

```diff
@@ -168,36 +151,67 @@ struct AssetScene: View {
                 )
                 .toolbar {
                     ToolbarItem(placement: .navigationBarLeading) {
-                        Button(Localized.Common.done) {
-                            isPresentingStaking.toggle()
-                        }
+                        Button(Localized.Common.done, action: onToggleStacking)
                         .bold()
                     }
                 }
                 .navigationBarTitleDisplayMode(.inline)
             }
         }
-        .taskOnce {
-            Task { try await fetch() }
-            Task { await model.updateAsset() }
-        }
+        .taskOnce(onTaskOnce)
         .navigationTitle(model.title)
     }
-    
-    func fetch() async throws {
-        async let updateAsset: () = try walletService.updateAsset(
-            wallet: model.walletModel.wallet,
-            assetId: model.assetModel.asset.id
-        )
-        async let updateTransactions: () = try model.fetchTransactions()
-        let _ = try await [updateAsset, updateTransactions]
+}
+
+// MARK: - Actions
+
+extension AssetScene {
+    @MainActor
+    private func onSelectHeader(_ buttonType: HeaderButtonType) {
+        isPresentingAssetSelectType = SelectAssetInput(type: buttonType.selectType, assetAddress: assetData.assetAddress)
+    }
+
+    @MainActor
+    private func onToggleStacking() {
+        isPresentingStaking.toggle()
+    }
+
+    @MainActor
+    private func onOpenLink(_ url: URL) {
+        // TODO: - find issue why we can't use @Environment(\.openURL) private var openURL here
+        // once we add native env url openning, we go to recursion on NavigationLink(value: model.assetModel.asset) { }
+        // AssetSceneViewModel recreates infinitly
+        guard UIApplication.shared.canOpenURL(url) else { return }
+        UIApplication.shared.open(url)
+    }
+
+    @MainActor
+    private func onSelectOptions() {
+        showingOptions = true
+    }
+
+    private func onTaskOnce() {
+        fetch()
+        Task { await model.updateAsset() }
     }
 }
 
-//struct AssetScene_Previews: PreviewProvider {
-//    static var previews: some View {
-//        AssetScene(
-//            model: .from(walletModel: WalletViewModel(wallet: .main), assetData: .main)
-//        )
-//    }
-//}
+// MARK: - Effects
+
+extension AssetScene {
+    private func fetch() {
+        Task {
+            do {
```

Comment:
can you wrap this method into view model? so you just call `model.update(wallet)`